pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs

```diff
@@ -2654,48 +2654,57 @@ public class PeptideTipProvider : ITipProvider
+                    _pepInfo.GetPeptideInfo(_matcher, out _settings, out transitionGroup, out mods);
+                    // build seq parts to draw
+                    _seqPartsToDraw = GetSequencePartsToDraw(mods);
+                    // Get small molecule info if any
+                    _smallMoleculePartsToDraw = null;
+                    var smallMolInfo = _pepInfo.GetSmallMoleculeLibraryAttributes();
+                    if (smallMolInfo != null && !smallMolInfo.IsEmpty)
+                    {
+                        // Get a list of things like Name:caffeine, Formula:C8H10N4O2, InChIKey:RYYVLZVUVIJVGH-UHFFFAOYSA-N MoleculeIds: CAS:58-08-2\tKEGG:D00528
+                        _smallMoleculePartsToDraw = smallMolInfo.LocalizedKeyValuePairs;
+                    }
 
-                if (_pepInfo.Target != null)
-                {
-                    // build mz range parts to draw
-                    _mz = _pepInfo.CalcMz(_settings, transitionGroup, mods);
-                    _mzRangePartsToDraw = GetMzRangeItemsToDraw(_mz);
-                }
-                else
-                {
-                    _mzRangePartsToDraw = new List<TextColor>();
-                    var precursorKey = _pepInfo.Key.LibraryKey as PrecursorLibraryKey;
-                    if (precursorKey != null)
+                    if (_pepInfo.Target != null)
                     {
-                        _mz = precursorKey.Mz;
+                        // build mz range parts to draw
+                        _mz = _pepInfo.CalcMz(_settings, transitionGroup, mods);
+                        _mzRangePartsToDraw = GetMzRangeItemsToDraw(_mz);
+                    }
+                    else
+                    {
+                        _mzRangePartsToDraw = new List<TextColor>();
+                        var precursorKey = _pepInfo.Key.LibraryKey as PrecursorLibraryKey;
+                        if (precursorKey != null)
+                        {
+                            _mz = precursorKey.Mz;
+                        }
                     }
-                }
 
-                // Ion mobility
-                var bestSpectrum = selectedLibrary.GetSpectra(_pepInfo.Key,
-                    IsotopeLabelType.light, LibraryRedundancy.best).FirstOrDefault();
-                if (bestSpectrum != null)
+                    // Ion mobility
+                    var bestSpectrum = selectedLibrary.GetSpectra(_pepInfo.Key,
+                        IsotopeLabelType.light, LibraryRedundancy.best).FirstOrDefault();
+                    if (bestSpectrum != null)
+                    {
+                        _ionMobility = bestSpectrum.IonMobilityInfo;
+                    }
+                }
+                catch (Exception e)
                 {
-                    _ionMobility = bestSpectrum.IonMobilityInfo;
+                    _errorMessage = e.Message; // Very annoying to encounter a blocking window in a mouseover, just show error in the tip

```

Comment:
Should check whether the exception is a programming error. I really prefer much more targeted exception handling than this kind of bulk handling over the Exception type. Also, I would prefer to see you introduce a wrapper function with the try-catch in it rather than just indenting the entire contents of a function into a try-catch block. Using a new wrapper will produce easier-to-understand diffs.

Reply:
Yes, should have used IsProgrammingDefect, that's in there now.

I'm not sure what you're asking for re wrappers though. The diff is already pretty simple, especially if your diff tool can ignore indent changes. If you want to wrap the ctor where it's invoked, that's in more than one place so the diff gets even bigger. If you want to move the guts of the ctor to a sub function that doesn't seem any less complicated to understand in a diff than just indenting. But I'm happy to do whatever you want once I understand it.

Reply:
By "wrapper", I mean that you would create a new method called something like "void Init(...)" that gets called inside the constructor, which becomes just:
try
{
Init();
}
Catch (Exception e)
{
}

A diff tool will do much better with this kind of change, and it avoids simply indenting everything because you want to add exception handling.

Reply:
Still seems like a more violent change to me - when you move the guts of the ctor try block elsewhere you get this:

![image](https://github.com/ProteoWizard/pwiz/assets/1200761/0b6f3142-ebb9-4593-abfd-1f8de5163193)

This strikes me as even more distracting. But again, happy to do so if that's your preference.

If this ctor had been written with exception handling in the first place, and with the form it has now, would you have felt moving most of the ctor code to a subfunction necessary or even desirable?

Reply:
Strict believer in non-violence that I am, I will concede. I didn't originally notice it was a constructor when I made the suggestion. That does make things more complicated. I have used the pattern I described with functions needing exception handling quite successfully in the past, but a constructor with readonly variables makes life a bit more complicated.

I suppose you could create another constructor with different variables, e.g. make assignments that can't throw, and then call an inner constructor with fewer arguments wrapped in a try-catch. probably not worth it.

I just don't love seeing this kind of change in the code. Oh well. At least it now can avoid catching programming errors somewhat.